File: TS/aqi.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var dict = (function () {
    function dict(keys, vals) {
        this.keys = keys;
        this.vals = vals;
        if (this.keys.length !== this.vals.length) {
            throw Error('The number of keys is not equal to the number of values. This dict is invalid.');
        }
    }
    dict.prototype.getval = function (key, otherwise) {
        if (otherwise === void 0) { otherwise = undefined; }
        var x = this.keys.indexOf(key);
        if (x === -1) {
            return otherwise;
        }
        else {
            return this.vals[x];
        }
    };
    dict.prototype.update = function (key, val) {
        var x = this.keys.indexOf(key);
        if (x === -1) {
            this.keys.push(key);
            this.vals.push(val);
        }
        else {
            this.vals[x] = val;
        }
    };
    return dict;
})();
exports.dict = dict;
var Aqi = (function () {
    function Aqi(pdict) {
        this.vals = pdict;
        this.des = '';
    }
    return Aqi;
})();
exports.Aqi = Aqi;
var Eaqi = (function (_super) {
    __extends(Eaqi, _super);
    function Eaqi(pdict) {
        _super.call(this, pdict);
        this.EAQI = new dict(['pm2', 'pm10', 'no2', 'o3', 'so2'], [[10, 20, 25, 50, 75, 800], [20, 40, 50, 100, 150, 1200], [40, 90, 120, 230, 340, 1000], [50, 100, 130, 240, 380, 800], [100, 200, 350, 500, 750, 1250]]);
        this.IDX = new dict([1, 2, 3, 4, 5, 6], ['Good', 'Fair', 'Moderate', 'Poor', 'Very Poor', 'Extremely Poor']);
        this.DES = new dict([1, 2, 3, 4, 5, 6], [['The air quality is good. Enjoy your usual outdoor activities.', 'The air quality is good. Enjoy your usual outdoor activities.'], ['Enjoy your usual outdoor activities.', 'Enjoy your usual outdoor activities.'], ['Enjoy your usual outdoor activities.', 'Consider reducing intense outdoor activities, if you experience symptoms.'], ['Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Consider reducing physical activities, particularly outdoors, especially if you experience symptoms.'], ['Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Reduce physical activities, particularly outdoors, especially if you experience symptoms.'], ['Reduce physical activities outdoors.', 'Avoid physical activities outdoors.']]);
        this.colour = new dict([1, 2, 3, 4, 5, 6], ['#0000ff', '#00cc99', '#ffff00', '#f75133', '#800000', '#800080']);
    }
    Eaqi.prototype.setres = function () {
        var caqi = [];
        for (var i = 0; i < this.vals.keys.length; i++) {
            var x = this.vals.keys[i];
            var thresh = this.EAQI.getval(x);
            if (this.vals.getval(x) > thresh[5]) {
                this.des = 'How are you still alive?';
            }
            var j = 0;
            for (var k = 0; k < thresh.length; k++) {
                j = thresh[k];
                if (this.vals.getval(x) <= j) {
                    break;
                }
            }
            var y = thresh.indexOf(j) + 1;
            caqi.push(y);
        }
        this.idx = Math.max.apply(Math, caqi);
        this.res = this.IDX.getval(this.idx, '');
    };
    Eaqi.prototype.setdes = function () {
        if (this.des === '') {
            this.des = "Healthy Individuals: " + this.DES.getval(this.idx, ['Invalid', 'Invalid'])[0] + " \nIndividuals with pre-existing conditions: " + this.DES.getval(this.idx, ['Invalid', 'Invalid'])[1];
        }
    };
    Eaqi.prototype.setcol = function () {
        if (this.des === 'How are you still alive?') {
            this.col = '#000000';
        }
        else {
            this.col = this.colour.getval(this.idx, '#ffffff');
        }
    };
    return Eaqi;
})(Aqi);
exports.Eaqi = Eaqi;
var Naqi = (function (_super) {
    __extends(Naqi, _super);
    function Naqi(pdict) {
        _super.call(this, pdict);
        try {
            if (this.vals.getval('co') === undefined) {
                throw Error('key co does not exist');
            }
            else {
                this.vals.update('co', (this.vals.getval('co') / 100));
            }
        }
        catch (error) { }
        this.NAQI = new dict(['pm2', 'pm10', 'no2', 'o3', 'so2', 'co'], [[30, 60, 90, 120, 250], [50, 100, 250, 350, 430], [40, 80, 180, 280, 400], [50, 100, 168, 208, 748], [40, 80, 380, 800, 1600], [10, 20, 100, 170, 340]]);
        this.DES = new dict([1, 2, 3, 4, 5, 6], ['Good', 'Satisfactory', 'Moderate', 'Poor', 'Very Poor', 'Severe']);
        this.Ival = new dict([1, 2, 3, 4, 5], [[0, 50], [51, 100], [101, 200], [201, 300], [301, 400]]);
        this.colour = new dict([1, 2, 3, 4, 5, 6], ['#009933', '#58ff09', '#ffff00', '#ffa500', '#ff0000', '#990000']);
    }
    Naqi.prototype.setres = function () {
        var caqi = [];
        var ind = [];
        for (var i = 0; i < this.vals.keys.length; i++) {
            var x = this.vals.keys[i];
            var thresh = this.NAQI.getval(x);
            if (this.vals.getval(i) > thresh[4]) {
                caqi.push(Math.round((((401 / thresh[4]) * (this.vals.getval(i) - thresh[4])) + 401)));
                ind.push(6);
            }
            else {
                var j = 0;
                for (var k = 0; k < thresh.length; k++) {
                    j = thresh[k];
                    if (this.vals.getval(x) <= j) {
                        break;
                    }
                }
                var y = thresh.indexOf(j);
                ind.push();
                if (y !== 0) {
                    if (this.vals.getval(x) < thresh[y - 1] + 1) {
                        j = thresh[y - 1];
                        y = thresh.indexOf(j);
                    }
                }
                var z = y + 1;
                var _a = this.Ival.getval(z, [0, 0]), Il = _a[0], Ih = _a[1];
                var Bl = 0;
                if (y !== 0) {
                    Bl = thresh[y - 1] + 1;
                }
                var Bh = j;
                caqi.push(Math.round((((Ih - Il) / (Bh - Bl)) * (this.vals.getval(x) - Bl)) + Il));
            }
        }
        this.res = Math.max.apply(Math, caqi).toString();
        this.idx = Math.max.apply(Math, ind);
    };
    return Naqi;
})(Aqi);
exports.Naqi = Naqi;

File: source/aqi.py
# AQICALC for the calculation of air quality index
#     Copyright (C) 2020  Varun Jain , Saadi Save

#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU Affero General Public License as published
#     by the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.

#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU Affero General Public License for more details.

#     You should have received a copy of the GNU Affero General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.

from typing import Dict, List, Tuple, Union


class Aqi:
    def __init__(self, pdict: Dict[str, float]) -> None:
        for i in pdict:
            if i not in ["pm2", "pm10", "no2", "o3", "so2", "co"]:
                raise ValueError("invalid pollutant used")
        self.vals = pdict
        self.res: str = ''
        self.des: str = ''
        self.col: str = ''


class Eaqi(Aqi):
    def __init__(self, pdict) -> None:
        Aqi.__init__(self, pdict)
        try:
            self.vals.pop('co')
        except KeyError:
            pass
        if not self.vals:
            raise ValueError("pdict should contain atleast one other pollutant than 'co' for Eaqi")
        self.__EAQI: Dict[str, List[int]] = {
            'AQI': [1, 2, 3, 4, 5, 6],
            'pm2': [10, 20, 25, 50, 75, 800],
            'pm10': [20, 40, 50, 100, 150, 1200],
            'no2': [40, 90, 120, 230, 340, 1000],
            'o3': [50, 100, 130, 240, 380, 800],
            'so2': [100, 200, 350, 500, 750, 1250]
        }
        self.__IDX: Dict[int, str] = {
            1: 'Good',
            2: 'Fair',
            3: 'Moderate',
            4: 'Poor',
            5: 'Very Poor',
            6: 'Extremely Poor'
        }
        self.__idx: int = 0
        self.__DES: Dict[int, Tuple[str, str]] = {
            1: ('The air quality is good. Enjoy your usual outdoor activities.', 'The air quality is good. Enjoy your usual outdoor activities.'),
            2: ('Enjoy your usual outdoor activities.', 'Enjoy your usual outdoor activities.'),
            3: ('Enjoy your usual outdoor activities.', 'Consider reducing intense outdoor activities, if you experience symptoms.'),
            4: ('Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Consider reducing physical activities, particularly outdoors, especially if you experience symptoms.'),
            5: ('Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Reduce physical activities, particularly outdoors, especially if you experience symptoms.'),
            6: ('Reduce physical activities outdoors.', 'Avoid physical activities outdoors.')
        }
        self.__colour: Dict[int, str] = {
            1: '#00ffff',
            2: '#00cc99',
            3: '#ffff00',
            4: '#f75133',
            5: '#800000',
            6: '#800080'
        }

    def set_res(self):
        caqi: List[int] = []
        for i in self.vals:
            thresh = self.__EAQI.get(i)
            if self.vals.get(i) > thresh[5]:
                self.des = 'How are you even alive?'
            j = 0
            for j in thresh:
                if self.vals.get(i) <= j: break
            caqi.append(thresh.index(j) + 1)
        self.__idx = max(caqi)
        self.res = self.__IDX.get(self.__idx, '')

    def set_des(self):
        if self.des == '':
            a, b = self.__DES.get(self.__idx, ('Invalid', 'Invalid'))
            self.des = f'Health messages:\nGeneral population: {a}\nSensitive populations: {b}'

    def set_col(self):
        if self.des == 'How are you even alive?':
            self.col = '#000000'
        else:
            self.col = self.__colour.get(self.__idx, '#ffffff')


class Naqi(Aqi):
    def __init__(self, pdict) -> None:
        Aqi.__init__(self, pdict)
        try:
            self.vals['co'] /= 100
        except KeyError:
            pass
        self.__NAQI: Dict[str, List[int]] = {
            'AQI': [1, 2, 3, 4, 5],
            'pm2': [30, 60, 90, 120, 250],
            'pm10': [50, 100, 250, 350, 430],
            'no2': [40, 80, 180, 280, 400],
            'o3': [50, 100, 168, 208, 748],
            'so2': [40, 80, 380, 800, 1600],
            'co': [10, 20, 100, 170, 340]
        }
        self.__idx: int = 0
        self.__DES: Dict[int, str] = {
            1: 'Good',
            2: 'Satisfactory',
            3: 'Moderate',
            4: 'Poor',
            5: 'Very Poor',
            6: 'Severe'
        }
        self.__HM: Dict[int, str] = {
            1: 'Minimal Impact',
            2: 'May cause minor breathing discomfort to sensitive people',
            3: 'May cause breathing discomfort to the people with lung disease such as asthma and discomfort to people with heart disease, children and older adults',
            4: 'May cause breathing discomfort to people on prolonged exposure and discomfort to people with heart disease with short exposure',
            5: 'May cause respiratory illness to the people on prolonged exposure. Effects may be more pronounced in people with lung and heart diseases',
            6: 'May cause respiratory effects even on healthy people and serious health impacts on people with lung/heart diseases. The health impacts may be experienced even during light physical activity'
        }
        self.__Ival: Dict[int, Tuple[int, int]] = {
            1: (0, 50),
            2: (51, 100),
            3: (101, 200),
            4: (201, 300),
            5: (301, 400)
        }
        self.__colour: Dict[int, str] = {
            1: '#009933',
            2: '#58ff09',
            3: '#ffff00',
            4: '#ffa500',
            5: '#ff0000',
            6: '#990000'
        }

    def set_res(self):
        caqi: List[int] = []
        idx: List[int] = []
        for i in self.vals:
            thresh = self.__NAQI.get(i)
            if self.vals.get(i) > thresh[4]:
                caqi.append(int(round((((401 / thresh[4]) * (self.vals.get(i) - thresh[4])) + 401), 0)))
                idx.append(6)
            else:
                j = 0
                for j in thresh:
                    if self.vals.get(i) <= j: break
                idx.append(thresh.index(j) + 1)
                if thresh.index(j) != 0:
                    if self.vals.get(i) < thresh[thresh.index(j) - 1] + 1:
                        j = thresh[thresh.index(j) - 1]
                x = thresh.index(j) + 1
                Il, Ih = self.__Ival.get(x, (0, 0))
                if thresh.index(j) != 0:
                    Bl = thresh[thresh.index(j) - 1] + 1
                else:
                    Bl = 0
                Bh = j
                caqi.append(int(round(((((Ih - Il) / (Bh - Bl)) * (self.vals.get(i) - Bl)) + Il), 0)))

        self.res = str(max(caqi))
        self.__idx = max(idx)

    def set_des(self):
        if int(self.res) > 500:
            self.des = 'Extremely severe. Avoid going outdoors.'
        else:
            self.des = f'Description: {self.__DES.get(self.__idx)}\nHealth message: {self.__HM.get(self.__idx)}'

    def set_col(self):
        if self.des == 'Extremely severe. Avoid going outdoors.':
            self.col = '#000000'
        else:
            self.col = self.__colour.get(self.__idx, '#ffffff')

class Mmaqi(Aqi):
    def __init__(self, pdict) -> None:
        Aqi.__init__(self, pdict)
        try:
            self.vals['co'] /= 100
        except KeyError:
            pass
        self.__MmAQI: Dict[str, List[int]] = {
            'AQI': [1, 2, 3, 4, 5],
            'pm2': [10, 25, 50, 75, 150],
            'pm10': [20, 50, 75, 150, 230],
            'no2': [40, 80, 120, 230, 340],
            'o3': [50, 100, 130, 240, 380],
            'so2': [40, 80, 200, 500, 750],
            'co' : [30, 100, 150, 250, 340]
        }
        self.__idx: int = 0
        self.__DES: Dict[int, str] = {
            1: 'Good',
            2: 'Fair',
            3: 'Moderate',
            4: 'Poor',
            5: 'Very Poor',
            6: 'Extremely poor',
            7: 'Severe',
            8: 'Hazardous'
        }
        self.__HM: Dict[int, Tuple[str, str]] = {
            1: ('The air quality is good. Enjoy your usual outdoor activities.', 'The air quality is good. Enjoy your usual outdoor activities.'),
            2: ('Enjoy your usual outdoor activities.', 'Enjoy your usual outdoor activities.'),
            3: ('Enjoy your usual outdoor activities.', 'Consider reducing intense outdoor activities, if you experience symptoms.'),
            4: ('Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Consider reducing physical activities, particularly outdoors, especially if you experience symptoms.'),
            5: ('Consider reducing intense activities outdoors, if you experience symptoms such as sore eyes, a cough or sore throat.', 'Reduce physical activities, particularly outdoors, especially if you experience symptoms.'),
            6: ('Reduce physical activities outdoors.', 'Avoid physical activities outdoors.'),
            7: ('Avoid physical activities outdoors.', 'Do not go outdoors'),
            8: ('Do not go outdoors. An air purifier is recommended.', 'Avoid coming into contact with any outdoor pollution. An air purifier is essential.')
        }
        self.__Ival: Dict[int, Tuple[int, int]] = {
            1: (0, 50),
            2: (51, 100),
            3: (101, 200),
            4: (201, 300),
            5: (301, 400)
        }
        self.__colour: Dict[int, str] = {
            1: '#00ffff',
            2: '#00cc99',
            3: '#ffff00',
            4: '#f75133',
            5: '#800000',
            6: '#800080',
            7: '#000000',
            8: '#000000'
        }

    def set_res(self):
        caqi: List[int] = []
        idx: List[int] = []
        for i in self.vals:
            thresh = self.__MmAQI.get(i)
            if self.vals.get(i) > thresh[4]:
                caqi.append(int(round((((401 / thresh[4]) * (self.vals.get(i) - thresh[4])) + 401), 0)))
                idx.append(6)
            else:
                j = 0
                for j in thresh:
                    if self.vals.get(i) <= j: break
                idx.append(thresh.index(j) + 1)
                if thresh.index(j) != 0:
                    if self.vals.get(i) < thresh[thresh.index(j) - 1] + 1:
                        j = thresh[thresh.index(j) - 1]
                x = thresh.index(j) + 1
                Il, Ih = self.__Ival.get(x, (0, 0))
                if thresh.index(j) != 0:
                    Bl = thresh[thresh.index(j) - 1] + 1
                else:
                    Bl = 0
                Bh = j
                caqi.append(int(round(((((Ih - Il) / (Bh - Bl)) * (self.vals.get(i) - Bl)) + Il), 0)))

        self.res = str(max(caqi))
        self.__idx = max(idx)
        if (int(self.res) > 500 and int(self.res) < 775):
            self.__idx = 7
        elif int(self.res) >= 775:
            self.__idx = 8

    def set_des(self):
        a, b = self.__HM.get(self.__idx, ('Invalid', 'Invalid'))
        self.des = f'{self.__DES.get(self.__idx)}\nHealth messages:\nGeneral population: {a} \nSensitive populations: {b}'

    def set_col(self):
        self.col = self.__colour.get(self.__idx, '#ffffff')


def convert(pollutant: str, value: float, unit: str):
    y = {
        'CO': 28,
        'NO2': 46,
        'O3': 48,
        'SO2': 64
    }
    if unit == 'ppm':
        mass = y[pollutant]
        value *= 0.0409 * mass * 1000
    elif unit == 'ppb':
        mass = y[pollutant]
        value *= 0.0409 * mass
    return round(value, 2)


def compare(a: Tuple[int, int], b: Tuple[int, int]) -> Union[int, str]:
    """a is the index to be compared with b
    \nformat:
    \ncompare(a=(a, amax), b=(b, bmax))
    """
    a1, a2 = a
    b1, b2 = b
    return int(round((1 - ((b1 * a2) / (a1 * b2))) * 100, 0))